src/Build/Graph/GraphBuilder.cs

```diff
@@ -0,0 +1,397 @@
+                            {
+                                // we have reached the nth project in the cycle, form error message and throw
+                                loadReference.projectsInCycle.Add(referenceConfig.ProjectFullPath);
+                                loadReference.projectsInCycle.Add(node.ProjectInstance.FullPath);
+                                var errorMessage = FormatCircularDependencyError(loadReference.projectsInCycle);
+                                throw new CircularDependencyException(
+                                    string.Format(
+                                        ResourceUtilities.GetResourceString("CircularDependencyInProjectGraph"),
+                                        errorMessage));
+                            }
+                            // this is one of the projects in the circular dependency
+                            // update the list of projects in cycle and return the list to the caller
+                            loadReference.projectsInCycle.Add(referenceConfig.ProjectFullPath);
+                            return (false, loadReference.projectsInCycle);
+                        }
+                    }
+
+                    node.AddProjectReference(referenceNode, projectReferenceItem, edges);
+                }
+
+                nodeState[node] = NodeVisitationState.Processed;
+                return (true, null);
+            }
+        }
+
+        private ProjectGraphNode CreateNewNode(
+            ConfigurationMetadata configurationMetadata,
+            ProjectCollection projectCollection,
+            ProjectGraph.ProjectInstanceFactoryFunc projectInstanceFactory,
+            ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode> allParsedProjects)
+        {
+            // TODO: ProjectInstance just converts the dictionary back to a PropertyDictionary, so find a way to directly provide it.
+            var globalProperties = configurationMetadata.GlobalProperties.ToDictionary();
+
+            var projectInstance = projectInstanceFactory(
+                configurationMetadata.ProjectFullPath,
+                globalProperties,
+                projectCollection);
+            if (projectInstance == null)
+            {
+                throw new InvalidOperationException(ResourceUtilities.GetResourceString("NullReferenceFromProjectInstanceFactory"));
+            }
+
+            var graphNode = new ProjectGraphNode(
+                projectInstance);
+            allParsedProjects[configurationMetadata] = graphNode;
+            return graphNode;
+        }
+
+        /// <summary>
+        ///     Load a graph with root node at entryProjectFile
+        ///     Maintain a queue of projects to be processed and evaluate projects in parallel
+        ///     Returns false if loading the graph is not successful
+        /// </summary>
+        private bool FindGraphNodes(
```

Comment:
Should this be LoadGraphNodes instead? Find seems to imply a search and I would not expect modifications.

Reply:
It finds and loads them, but does not connect them. What do you think about `FindAndLoadGraphNodes`? Seems a bit superfluous.